Prune unused artifacts from non-static builds

henriquesimoes opened the pull request:
Both `dynamic-link` and `no-build` targets generate unnecessarily large IOC images due to the trivial `COPY` done from `/opt`. Clean up everything from there and `/usr/local` directory before doing the `COPY` to the IOC image.

Clean up strategy proposed here is the following:

- Remove entirely any module that does not have a shared library linked against binaries under the IOC target directories (i.e. `/opt/$REPONAME` or `$RUNDIR`);
- Prune directories from used modules and EPICS base not containing shared libraries, EPICS databases or autosave requirement files;
- Remove any shared library not linked against binaries under IOC target directories;
- Remove all statically linked libraries;

For pruning module directories, another approach that could be followed is to delegate this responsibility to `install_module`. This way, we can have contextualized pruning for modules. On the other hand, this cannot be as aggressive as implemented in this proposal, as build-dependent files (the ones defining the build-system itself, for instance) must exist until the IOC build is complete.

I have tested this with [ad-aravis-epics-ioc](https://github.com/cnpem/ad-aravis-epics-ioc), and it shrinks the image size from 1.33GB down to 322MB.

This should be further tested and extensively reviewed, so that no artifact is removed unexpectedly.

henriquesimoes commented on base/Dockerfile:
```diff
@@ -66,3 +66,5 @@ RUN ./install_area_detector.sh
 COPY install_motor.sh .
 RUN ./install_motor.sh
+
+COPY lnls-prune-artifacts.sh /usr/local/bin/lnls-prune-artifacts
```
I've added this last so that image cache is not invalided everything for nothing. It can be moved to be side-by-side with `lnls-run` after the script ready to be merged.

henriquesimoes replied:
Made this a permanent change now. ;)

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,122 @@
+
+    modules="$(echo "$release_defs" | grep $EPICS_MODULES_PATH)"
+
+    for module in $modules; do
+        echo $module
+    done
```
Doing this instead of simply echoing out `"$modules"` removes empty lines. Maybe there is a shorter and cleaner way to do this.

ericonr replied:
I'd use the sed answer from <https://serverfault.com/questions/252921/how-to-remove-empty-blank-lines-from-a-file-in-unix-including-spaces>

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,122 @@
+        if [ -n "$libs" ]; then
+            size=$(du -hsc $libs | tail -n 1 | cut -f 1)
+
+            echo "Removing static libraries from $target ($size)"
```
I'm collapsing all static libraries. Should we list the removed files instead?

I decided to let it collapsed because it is simpler to implement and no application should need them after build.

ericonr replied:
I'm okay with them being collapsed.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,122 @@
+
+remove_static_libs() {
+    for target; do
+        libs=$(find $target -type f -name *.a -print)
```
I'm assuming there are only [AR archives](https://en.wikipedia.org/wiki/Ar_(Unix)) which are statically linked libraries. Is that a really good assumption?

ericonr replied:
For running on Linux, I think it should be.

If you want to have a sort of lint for our build process, after the base image is done building stuff, I'd run a script to make sure we don't have any `.o` files leftover. They are the only other kind of "static" files we could have.

henriquesimoes replied:
I've checked our base image, and it does not have any `*.o` left behind. Our cleaning done after installation seems to be enough for now to remove these.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,122 @@
+    remove_libs=$(find_shared_libs /opt /usr/local)
+    used_libs=$(get_linked_libraries $targets)
+
+    for lib in $used_libs; do
+        remove_libs=$(echo "$remove_libs" | grep -v $lib)
+    done
+
+    for lib in $remove_libs; do
+        size=$(du -hs $lib | cut -f 1)
+
+        echo "Removing shared library '$lib' ($size)"
+        rm -f $lib
+    done
```
I don't know if there is an easier way to do the set difference besides looping twice while `grep`ing out.

ericonr replied:
> I don't know if there is an easier way to do the set difference

This was the part that made me question if python wouldn't be best. (and calling `ldd` as a subprocess in python would be ok for me, with the added advantage of python allowing more expressiveness in text processing).

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,122 @@
+        $(find $module -type f -regex ".*\.\(db\|template\|req\)" -printf "%h\n" | sort | uniq)
+    "
+
+    for candidate in $(find $module -type d); do
+        [ -d $candidate ] || continue
+
+        if [[ ! $keep_paths =~ "$candidate".* ]]; then
+            size=$(du -hs $candidate | cut -f 1)
+
+            printf "Removing directory '$candidate' ($size)...\n"
+            rm -rf $candidate
+        fi
```
In this loop, I'm removing only directories as a whole if no important file (listed in `keep_paths`) is stored in it.

Another approach is to do this in a file basis, which could prune modules even further. I'm not sure if that's the best way to handle it, though.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,122 @@
+
+            echo "Removing module '$module' ($size)..."
+            rm -rf $module
+        elif [ -d $module/lib ]; then
+            prune_module_dirs $module
+        fi
```
I'm using the `lib` subdirectory to ignore top-level repositories listed in `RELEASE`, such as areaDetector. Is there a better approach?

ericonr replied:
I can think of hardcoding things, but that doesn't really seem better. Will think on it.

henriquesimoes replied:
Such hack is no longer needed with the latest rewrite. I'm filtering out target directories (and their parents), so we shouldn't call for top-level repos.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,122 @@
+    prune_module_dirs $EPICS_BASE_PATH
+}
+
+clean_up_epics_modules $target_paths
+remove_static_libs /opt
+remove_unused_shared_libs $target_paths
```
I'm not sure if `remove_static_libs` should receive an argument or not. Their companion functions receive directories which contain important binaries whose dependencies must be preserved. On the other hand, `remove_static_libs` currently receives the directories to prune the libraries off.

Maybe both of them should take `/opt` (and `/usr/local`?) from a shared constant variable instead.

Opinions on that?

henriquesimoes reviewed (commented):
I've added comments to points that I wasn't sure what is the best way to deal with, so that we can discuss if there are relevant or not and how to improve, if that's the case.

In general, I feel like I could have written helper functions that could be more easily shared among the steps. But that's not a concrete thing to propose anything. Let me know if you feel the same.

ericonr reviewed (commented):
The first two commits are correct and can be merged separately, if you'd like to split them.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+
+target_paths=$@
+
+
```
I don't think this variable needs to be set; the function call at the end can simply use `$@` itself.

Putting this at the start of the script makes it look like you are going to manipulate the array in the middle of it.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+get_used_epics_modules() {
+    targets=$@
+
+    executables=$(find $targets -type f -executable -print)
```
`-print` is the default, and I had to search for it to see if it changed anything and confirm it's the default. I'd rather omit it.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+    executables=$(find $targets -type f -executable -print)
+    libs="$(ldd $executables 2>/dev/null)"
+
+    modules=()
```
Leftover variable

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+    targets=$@
+
+    executables=$(find $targets -type f -executable -print)
+    libs="$(ldd $executables 2>/dev/null)"
```
`ldd` on musl doesn't check all arguments, it simply ignores all but the first one.

We should either explicitly loop or add a comment about depending on this behavior.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+    libs="$(ldd $executables 2>/dev/null)"
+
+    modules=()
+    epics_libs=$(echo "$libs" | grep -oE "$EPICS_MODULES_PATH/.*/*.so")
```
I don't think the EPICS build system supports versioned sonames, so using `.so` is fine. Maybe mention that in the commit message?

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+
+    modules=()
+    epics_libs=$(echo "$libs" | grep -oE "$EPICS_MODULES_PATH/.*/*.so")
+    for lib in $epics_libs; do
+        module=${lib%/lib/*/*lib*so}
```
Use `epics_lib` or even something other than `lib`. Otherwise the substitution gets confusing with the repeated `lib`, and it's confusing enough already.

Add a comment explaining the purpose of expression (an example would be enough).

henriquesimoes replied:
Let me know what you think about the comment.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+    for lib in $epics_libs; do
+        module=${lib%/lib/*/*lib*so}
+        echo $module
+    done | sort | uniq
```
`sort -u` so we don't need `uniq`

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+}
+
+get_all_epics_modules() {
+    release_defs=$(cat ${EPICS_RELEASE_FILE} | grep = | cut -d'=' -f 2)
```
`grep = ${EPICS_RELEASE_FILE}`

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+get_all_epics_modules() {
+    release_defs=$(cat ${EPICS_RELEASE_FILE} | grep = | cut -d'=' -f 2)
+
+    modules="$(echo "$release_defs" | grep $EPICS_MODULES_PATH)"
+
+    echo "$modules"
```
You could simply have the last line be `echo $release_defs | grep ...`

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+    used_modules=$(get_used_epics_modules $targets)
+
+    for module in $(get_all_epics_modules); do
+        if [[ -d $module && ! $used_modules =~ "$module" ]]; then
+            size=$(du -hs $module | cut -f 1)
```
Why is `-d $module` required? Shouldn't they always exist? Or is this covering some case like an AreaDetector or motor submodule being in the list after the super-repository has already been cleaned? I think this should be explained somewhere.

I find `=~` somewhat non obvious to parse. And it means "motor" won't be removed if "motorX" is in used modules, right? I think it might make sense to print the list into a file (in /tmp, which can be removed afterwards), so it's split by newlines, and use grep to match the *whole* line.

henriquesimoes replied:
> And it means "motor" won't be removed if "motorX" is in used modules

Because of such issues with prefix matching, I moved to checking the each level of the path at a time, with whole line matching instead. This turned out to be handy for the following patches as well.

Still, don't feel like my implementation explores all bash-ims it could.

ericonr reviewed (commented):
The Dockerfile parts of the third commit are ok.

---

Commit message nits:

It's not the caching that's the issue, it's simply how container images are layered:

```
The clean up phase must be executed in the build stage (before the final
copy), otherwise the COPY layer would still make the image size big.
This implies `no-build` target needs to copy from a pruned version of
the base image, which is a new stage.
```

henriquesimoes commented on images/docker-compose-mca.yml:
```diff
@@ -8,6 +8,6 @@ services:
         org.opencontainers.image.source: https://github.com/cnpem/epics-in-docker
       args:
-        REPONAME: mca
+        REPONAME: epics/modules/mca
```
To be honest, I don't like very much the idea of reusing `REPONAME`. It seems awkward to me. I'd rather have them specified by a variable whose name makes sense at all.

I did so because it was easier to test other things first. Let me know what you think.

ericonr replied:
That's fair. Do you think it would lead to special casing these modules in the image Dockerfile?

henriquesimoes replied:
> Do you think it would lead to special casing these modules in the image Dockerfile?

They kinda already have such special handling... all of them are no-build targets. They of course currently have almost the same (compose variables) API of other targets. I say "almost" because they diverged by not using REPONAME before, and now by not meaning "base name to where to clone the repository" but rather "location inside /opt where important artifacts live".

I'm not sure if I answered your question at all, but I do feel like we have always being dealing with a special case.

ericonr replied:
I guess I am essentially wondering if we shouldn't use some other name in REPONAME, since these images are a special case.

henriquesimoes replied:
I changed this to `APP_DIRS`. It should encode full paths now. It is named after a plural noun; after all, we can have multiple directories that we want to ensure are properly working. This should work as a palliative measure for spurious cases where we clean up too much. Any directory listed there will have its belonging module removed from pruning (both module-based and directory-based).

Maybe we could be even more explicit about its usage regarding "not pruning it too much and keeping its executable binaries working". Any opinion and suggestion about that?

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+
+    keep_dirs="$targets $used_modules"
+
+    for module in $(filter_out_dirs "$unused_modules" "$keep_dirs"); do
+        # if we already removed it because of its top-level repository or
+        # because it is an IOC, move on to the next.
+        [ ! -d $module ] && continue
+
+        size=$(du -hs $module | cut -f 1)
+
+        echo "Removing module '$module' ($size)..."
+        rm -rf $module
+    done
```
This loop will misbehave if we turn out having modules with empty spaces.

I kept this as a loop with such requirement because we mostly control such paths. Let me know what you think about this.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+find_elf_executables() {
+    targets=$@
+
+    while read -r executable; do
+        read -r -N 4 trailer < "$executable"
+
+        # Output only ELF binaries
+        if [ "$trailer" = $'\x7fELF' ]; then
+            echo $executable
+        fi
+    done < <(find $targets -type f -executable)
```
We must loop this way here because they might be file names with spaces. Indeed, MCA is such a representative of such scenario.

Because of that, I wonder if I haven't assumed to much elsewhere regarding file names.

Also, would this deserve a comment or is this clear enough?

henriquesimoes replied:
> Do you mean using this instead of for executable in $(find ...?

Exactly.

> You can just leave a comment on top of the loop that file names might have spaces.

Okay. Should I do the same for every other loop as well?

ericonr replied:
> I kept this as a loop with such requirement because we mostly control such paths. Let me know what you think about this.

I agree with this justification here, especially because it makes loops more complicated.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+    # pull those not found libraries
+    found="$(echo "$linked" | grep -v "not found")"
+
+    # Get their full path
+    libs=$(echo "$found"| cut -d'>' -f 2 | cut -d' ' -f 2)
```
I feel like there is a better way to do this. I though abort `grep -o`, but it would require a pattern for any library path, including soname handling. I'm not sure if that turns out to be better.

henriquesimoes replied:
> Why doesn't using one `cut -d' ' -f 3` command work?

Actually, it does. I was trying not to rely on spaces at first, but I turned out doing so in the end. I didn't notice it could be easily simplified. Thanks.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+remove_unused_epics_modules() {
+    targets=$@
+
+    unused_modules=$(get_all_epics_modules)
```
This variable has the wrong name.

henriquesimoes replied:
Oh, thanks. Refactoring left over mistake for sure. `all_modules` should do the job.

ericonr reviewed (changes requested):
One thing isn't clear to me. From targets, you are obtaining unused modules, to then obtain used modules, to finally obtain unused modules again, and remove them. Is that logic necessary, can we simplify it? If you already have these answers, you can add them to the commit message.

Commit message nit:

> EPICS base is assumed to be always needed,

always be needed

> It is assumed we are using glibc implementation

we are using glibc's ldd

For the second commit, since we are removing stuff *after* building, static libraries can always be removed, notjust for non-static builds. I think it's only the commit message that's wrong.

Nit:

> Static libraries are not used neither during compile time nor runtime for non-static builds.

Static libraries are not used during compile nor runtime.

Static libraries are used neither during compile nor runtime.

> All AR archives are assumed to be static libraries,

All files with .a extension are assumed to be static libraries,

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+    echo "$libs" | sort -u
+}
+
+get_used_epics_modules() {
+    epics_libs=$(find_linked_libraries $@)
+    modules=$(get_all_epics_modules)
```
Okay. I didn't take this into account when ordering the functions. Will reorder them. Thanks.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+    # We grep out not found libraries, since they cannot be kept if we don't
+    # know where they are.
+    #
+    # Final binary may be actually runnable, since rpath of another binary may
+    # pull those not found libraries
+    found="$(echo "$linked" | grep -v "not found")"
```
Yep. For `libopcua.so` that comes in OPCUA module, both EPICS libCom and libdbCore won't be found if we simply `ldd` it.

ericonr replied:
> EPICS libCom and libdbCore won't be found if we simply ldd it.

Oh, that's annoying. Is that the only case?

I'm wondering if we should make it clear it's due to OPCUA, and remove it once/if we move to the open source impl, or if it makes sense to keep for any other similar cases with proprietary libraries.

henriquesimoes replied:
> Is that the only case?

That's the only one that I spotted. Considering that EPICS modules and IOC by default relies on those custom rpath entries, I'm sure any other already-compiled library shipped into module is susceptible to this issue. This probably falls into what you mention of proprietary libraries in most cases.

If we were using system installation for EPICS base, that wouldn't have been an issue even for OPCUA.

ericonr replied:
I'm pending towards mentioning in a comment that the "not-found" case is due to OPCUA, then.

henriquesimoes replied:
Added to the commit message. Let me know if you think that's not enough.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+    for item in $exclude_list; do
+        dir=$item
+
+        while [ "$dir" != "/" ]; do
+            list=$(echo "$list" | grep -xv $dir)
+
+            dir=$(dirname $dir)
+        done
```
Yeah, I turned out to do so when testing to avoid infinite loops, but didn't get to a final approach on how to handle it.

Should we abort in such cases, or simply ignore that entry?

ericonr replied:
IMO abort, it's breaking an expected invariant.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,91 @@
+    list="$1"
+    exclude_list="$2"
+
+    for item in $exclude_list; do
+        dir=$item
```
> This should include a comment indicating that for the first iteration, dir might not actually be a directory.

Should we call it `path` instead?

> And why not use for dir in...?

Good question. None, I'm afraid. When expressing the logic, that was the natural way to implement it, and I didn't simplify it in the end. Thanks for the notice.

ericonr replied:
> Should we call it path instead?

I think that works well!

gustavosr8 commented:
nit from 13072def851f1ea2ff39b96aca02b86f840da9e6:
`s/clean-up/cleanup`
`s/clean up/cleanup`
`s/so that it contains a valid mapping of all modules/so it contains a valid mapping of all modules` 
`s/assumes the inspected binaries/assumes that the inspected binaries`
`s/It is assumed we are using glibc's/It is assumed that we are using glibc's`

gustavosr8 commented:
> Thanks, @gustavosr8. I've also made `remove_unused_shared_libs` respect libraries inside `APP_DIRS`. This allows users to specify the path where their `dlopen(3)`ed libraries live. Can you test this again with the pyDevSup-based IOC?

Now it seems to work! Thanks :)
Just a reminder to document the use of `APP_DIRS` ;)

gustavosr8 reviewed (changes requested):
Please add instructions to put `/opt/epics/modules/pyDevSup/python3.9` under `APP_DIRS` (or any analogous var) when building an IOC that uses the pyDevSup module.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env bash
+
+set -Eeu
```
Mention in commit message that you need `-E` to properly propagate errors.

ericonr commented on images/docker-compose-mca.yml:
```diff
@@ -8,6 +8,6 @@ services:
         org.opencontainers.image.source: https://github.com/cnpem/epics-in-docker
       args:
-        REPONAME: mca
+        APP_DIRS: /opt/epics/modules/mca
```
Also mention in commit message that REPONAME wasn't relevant for these images, and that they were moved to use `APP_DIRS` so as not to break them.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -106,5 +112,23 @@ remove_static_libs() {
+remove_unused_shared_libs() {
+    target_libs=$(find_shared_libs $@)
+    linked_libs=$(find_linked_libraries $@)
+    remove_libs=$(find_shared_libs /opt /usr/local)
```
Adding `/usr/local` here is due to #55, right? Might make sense to mention in the commit that it's future proofing, and doesn't actually change behavior for now.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -36,6 +36,12 @@ find_elf_executables() {
     done < <(find $targets -type f -executable)
 }
 
+find_shared_libs() {
```
Consistency: libraries vs libs

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -77,7 +77,27 @@ get_used_epics_modules() {
+
+    keep_paths="
+        $(find_shared_libs $module)
+        $(find $module -type f -regex ".*\.\(db\|template\|req\)" -printf "%h\n" | sort -u)
```
Some IOCs use `substitutions` at runtime, I think they should also be kept...

Should ACFs be kept as well?

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -77,7 +77,27 @@ get_used_epics_modules() {
+    while read -r candidate; do
+        [ -d $candidate ] || continue
+
+        if [[ ! $keep_paths =~ "$candidate".* ]]; then
+            size=$(du -hs $candidate | cut -f 1)
+
+            printf "Removing directory '$candidate' ($size)...\n"
+            rm -rf $candidate
+        fi
```
Is this nearly the same operation as `filter_out_paths`? Is it open-coded because you need to support spaces?

henriquesimoes replied:
Yes, and yes. However, regex matching suffers from what you said elsewhere: if a candidate folder is a prefix of a path to keep, this method decides to keep them. This means we might remove less than we could.

Maybe supporting file paths with spaces to `filter_out_paths` is a way to go. What do you think?

ericonr replied:
That would mean you need to use arrays instead of lists in strings, right?

henriquesimoes replied:
Don't think so. `filter_out_paths` already receives two newline-separated strings. Stop using `for` and redirect the list to the `while read; do ... done` stdin should be enough.

henriquesimoes replied:
Doing so made possible to extract yet another function to be used both when removing entire modules and pruning only specific directories. ;)

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -77,12 +77,36 @@ get_used_epics_modules() {
+# their .lnls-keep-paths defined entries as absolute paths.
+get_defined_paths_to_keep() {
+    for path; do
+        depth=$(echo $path | grep -o '/' | wc -l)
+
+        for i in $(seq 1 $depth); do
+            ancestor=$(echo $path | cut -d '/' -f 1-$i)
```
Why can't this be a `dirname` loop that terminates on `/`?

ericonr replied:
Could also break on `.` and `/`, so it still works for local directories.

henriquesimoes replied:
> I like the dirname loop because it's already used elsewhere, so it's a pattern that's easy to spot.

Yeah, I agree on that. I changed this locally. I'm testing if it is still working as expected before pushing.

> Could also break on . and /, so it still works for local directories.

Indeed. I'll defer that to the future if needed. As I said, it actually is not a big deal for now.

ericonr reviewed (commented):
Commit message nits:

"workaround" should be "work around", it's being used as a verb.

"a "import"" should use "an"

ericonr commented:
I'm force pushing into the PR, after rebasing it and fixing some of my own nits.

Original commit: 0a9c79e85dcc5344798a594583f2874c6f154000

henriquesimoes commented:
Changed the last implementation bits now. I learned something new during this round: calling a bash function outside a subshell does not create another variable scope. I was assuming it did until now. If I knew this from start, I would have defined several other variables explicitly as `local` just to avoid overwriting something by mistake in the future.

I haven't done so now, though, to reduce the amount of noise in range diffs. I've only done so now where it would be a bug otherwise. Let me know if you think it is interesting to refactor that before merging.

From my point of view, documentation is the last topic to be covered in this PR.

henriquesimoes commented on README.md:
```diff
@@ -51,8 +51,17 @@ services:
-dependencies will also be copied.
+`dynamic-link`.
+
+For non-static builds, an automatic pruning step is executed to make the IOC
+image size small by removing unused artifacts. This step preserves all ELF
+executables inside `RUNDIR` and the IOC repository, as well as their dependent
+EPICS modules based on the linkage information. Additional paths can be
+provided in `args` under `APP_DIRS` key to extend the list of where important
+applications and shared libraries are, including any `dlopen(3)`ed libraries.
+Pruning step can be skipped entirely by providing the `SKIP_PRUNE=1` argument,
+but this is highly discouraged as it will increase significantly the image
+size.
```
Do you think it is interesting to specify we also prune directories from modules in use?

I don't like the idea of replicating what types of files are kept, but it may be okay to say they are also pruned.

ericonr replied:
I think it should be mentioned that some directories (e.g. GUI ones) are assumed to be unused and removed as well.

henriquesimoes replied:
Done. Feel free to propose an enhanced wording ;)

ericonr commented on README.md:
```diff
@@ -51,8 +51,17 @@ services:
+EPICS modules based on the linkage information. Additional paths can be
+provided in `args` under `APP_DIRS` key to extend the list of where important
+applications and shared libraries are, including any `dlopen(3)`ed libraries.
+Pruning step can be skipped entirely by providing the `SKIP_PRUNE=1` argument,
+but this is highly discouraged as it will increase significantly the image
+size.
```
If the resulting image fails at runtime, a careful look at error messages might provide clues for directories to add to `APP_DIRS`; if that isn't possible, or if it's believed that the directories added to `APP_DIRS` should have been detected automatically, please open an issue. Please do so as well if the pruning step errors out during the build. In case it is necessary to build a working image before these issues can be fixed, the pruning step can be skipped entirely by setting the `SKIP_PRUNE=1` argument; note that this is highly discouraged, because it increases the image size significantly.

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,198 @@
+
+set -Eeu
+
+filter_out_paths() {
```
This function is missing a comment that its arguments should be split by newlines

henriquesimoes reviewed (commented):
Two nits left I've noticed now:

- clean up `/opt/epics-in-docker` (#88)
- clean up `lnls-*` scripts we don't need (#59)

Maybe the last one can be deferred to the #59 itself, even though `lnls-get-n-unpack` cannot be removed by `lnls-prune-artifacts`.

ericonr commented:
> Two nits left I've noticed now:
> 
>     * clean up `/opt/epics-in-docker` ([base: remove repetition of environment variables. #88](https://github.com/cnpem/epics-in-docker/pull/88))
> 
>     * clean up `lnls-*` scripts we don't need ([Prune unused artifacts from non-static builds #59](https://github.com/cnpem/epics-in-docker/pull/59))
> 
> 
> Maybe the last one can be deferred to the #59 itself, even though `lnls-get-n-unpack` cannot be removed by `lnls-prune-artifacts`.

is 59 there correct? that's this PR

I'm reconsidering cleaning up `/opt/epics-in-docker`. It's composed of small text files, and the info there is a pretty easy way to see what versions of everything were used to build a certain image.

henriquesimoes commented:
> is 59 there correct? that's this PR

Oops. It is not. It is #55.

> I'm reconsidering cleaning up /opt/epics-in-docker. It's composed of small text files, and the info there is a pretty easy way to see what versions of everything were used to build a certain image.

Makes sense. Based on the first point, maybe it is not worth to remove auxiliary scripts as well.

gustavosr8 commented:
> I'm reconsidering cleaning up `/opt/epics-in-docker`. It's composed of small text files, and the info there is a pretty easy way to see what versions of everything were used to build a certain image.

Maybe if we use `LABELS` to keep track of these infos (#24) we could clear this up. WDYT?

ericonr commented:
> > I'm reconsidering cleaning up `/opt/epics-in-docker`. It's composed of small text files, and the info there is a pretty easy way to see what versions of everything were used to build a certain image.
> 
> Maybe if we use `LABELS` to keep track of these infos (#24) we could clear this up. WDYT?

I think I'd be okay with both rather than removing the files, either way.

ericonr commented on base/Dockerfile:
```diff
@@ -35,7 +35,6 @@ RUN apt update -y && \
         ca-certificates
 
 COPY lnls-get-n-unpack.sh /usr/local/bin/lnls-get-n-unpack
-COPY lnls-run.sh /usr/local/bin/lnls-run
```
This commit message is wrong. This script isn't executed during any build step.

henriquesimoes replied:
Indeed. I was willing to write in the affirmative tense, but made the complement subset smaller than it actually is. Thanks.

henriquesimoes commented:
Something is off when building the opcua image: it removes [`lib/` from `opcua-module`](https://github.com/cnpem/epics-in-docker/pull/59/checks#step:8:448). I'll investigate that.

henriquesimoes commented:
Okay, I've found the issue: epics-modules/opcua#169. As discussed in person with @ericonr, I'll drop this assumption of libraries having the execution bit set. This will make we check the first bytes of all files inside a module to be pruned, but it should give us a more general approach to deal with any library installed (including the ones from non-EPICS modules).

For building `opcua-epics-ioc` in my local machine (5 times), pruning by looking only executable files took $1.71 \pm 0.05$ second while checking all files took $1.77 \pm 0.07$ second. For `ad-aravis-epics-ioc`, it took $5.1 \pm 0.2$ seconds (only exe) vs $5.25 \pm 0.09$ seconds (all files). In the end, It looks like that's basically for free.

henriquesimoes commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,101 @@
+    linked_libs=$(find_linked_libraries $@)
+    all_modules=$(get_all_epics_modules)
+
+    unused_modules=$(filter_out_paths "$all_modules" "$linked_libs")
+
+    filter_out_paths "$all_modules" "$unused_modules"
```
This logic is actually broken. If we get in `$unused_modules` something including `/opt/epics/modules/motor/modules/motorMotorSim`, for instance, the second `filter_out_paths` will also remove `/opt/epics/modules/motor`, since it is a prefix of the first one. However, we don't want such "prefixes" removal in this last step, as it should be only a pure and simple set difference.

I'll fix that.

henriquesimoes commented:
I've just checked a couple of IOCs:

- our four built-in images;
- ad-aravis-epics-ioc
- ad-sim-epics-ioc
- motor-sim-epics-ioc

My tests consisted in building the container themselves, and checking if binaries run at all. I haven't checked all the files removal logs carefully for all of them, though. For the ones I did, everything seemed to make sense.

One thing I noticed during my tests with ad-sim-epics-ioc is that all directories from modules inside areaDetector are pruned at once when the top-level repo is pruned. This makes inner modules pruning basically a no-op. It doesn't cause anything harmful for this IOC, though. It does mean `.lnls-keep-paths` but live in `areaDetector` top-level directory (I now think I should have added this to the commit message, BTW).

<details>
<summary>Logs of pruning areaDetector modules for ad-sim-epics-ioc</summary>

```
Pruning module '/opt/epics/modules/areaDetector'...
Removing directory '/opt/epics/modules/areaDetector/ADCSimDetector' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADnED' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADBruker' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPcoWin' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADRIXSCam' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADSupport/configure' (36K)...
Removing directory '/opt/epics/modules/areaDetector/ADSupport/supportApp' (94M)...
Removing directory '/opt/epics/modules/areaDetector/ADSupport/include' (1.2M)...
Removing directory '/opt/epics/modules/areaDetector/ADSupport/bin' (208K)...
Removing directory '/opt/epics/modules/areaDetector/ADUVC' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADProsilica' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ci' (16K)...
Removing directory '/opt/epics/modules/areaDetector/ADPluginBar' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPSL' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPhotonII' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADSpinnaker' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPixirad' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/pvaDriver' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPCO' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPICam' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/configure' (136K)...
Removing directory '/opt/epics/modules/areaDetector/documentation' (148K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/configure' (36K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/documentation' (652K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/include' (12K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/iocs/simDetectorIOC/iocBoot/iocSimDetector/autosave' (8.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/iocs/simDetectorIOC/iocBoot/iocHDF5Test/autosave' (8.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/iocs/simDetectorIOC/configure' (36K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/iocs/simDetectorIOC/simDetectorApp' (28K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/iocs/simDetectorNoIOC' (68K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/simDetectorApp/src' (68K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/simDetectorApp/op' (728K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/docs' (856K)...
Removing directory '/opt/epics/modules/areaDetector/ADSimDetector/dbd' (8.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPilatus' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPerkinElmer' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPointGrey' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/ci' (16K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/configure' (36K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/documentation' (6.8M)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/XML_schema' (52K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/cfg' (16K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/include' (1.3M)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/ADApp/ntndArrayConverterSrc' (40K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/ADApp/op' (16M)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/ADApp/pluginTests' (308K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/ADApp/pluginSrc' (2.4M)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/ADApp/ADSrc' (256K)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/docs' (5.5M)...
Removing directory '/opt/epics/modules/areaDetector/ADCore/dbd' (44K)...
Removing directory '/opt/epics/modules/areaDetector/ADCameraLink' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADRoper' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/aravisGigE' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADAndor3' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADAndor' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADEiger' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPluginEdge' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADFastCCD' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADCompVision' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADADSC' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADVimba' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADMythen' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADPvCam' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADmar345' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/firewireDCAM' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADURL' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADViewers' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/specsAnalyser' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADmarCCD' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADDexela' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADFireWireWin' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/docs' (412K)...
Removing directory '/opt/epics/modules/areaDetector/ADMerlin' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADQImaging' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/NDDriverStdArrays' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADLightField' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ffmpegViewer' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADMMPAD' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ffmpegServer' (4.0K)...
Removing directory '/opt/epics/modules/areaDetector/ADLambda' (4.0K)...
Pruning module '/opt/epics/modules/areaDetector/ADCore'...
Pruning module '/opt/epics/modules/areaDetector/ADSimDetector'...
Pruning module '/opt/epics/modules/areaDetector/ADSupport'...
```

</details>

ericonr commented on base/lnls-prune-artifacts.sh:
```diff
@@ -0,0 +1,113 @@
+# Both list are treated as newline-separated strings, and must consist
+# exclusively of absolute paths.
+filter_out_paths() {
+    list="$1"
+    exclude_list=$(echo "$2" | grep -v '^$')
```
I think if we accept a list with empty lines for second argument, we should try to do the same for the first one too.

henriquesimoes replied:
Fair enough. I've also included a mention in the docstring.

ericonr reviewed (changes requested):
Considering the principle of each commit keeping the repository fully working, would it be possible to have "ioc: don't prune pre-selected directories." before "ioc: prune module directories from non-static builds."? Something like implement `get_defined_paths_to_keep`, then add it to relevant modules, then implement module pruning already using it? Would that make sense and be worth it?

I agree that the message for "base: don't prune HDF5 filter plugins from ADSupport." should mention the need to have the file in AreaDetector

Commit message nits:

> The set difference is works by

s/is//

> It is also susceptible to be skipped during module directory pruning if a target specifies it.

It can also be skipped

gustavosr8 commented:
nit on 264776462a7bf02c531c3657a6e00d4fc06c2cc1 message:
'...to give users a immediate last resort...' -> '...to give users an immediate last resort...'

henriquesimoes commented:
Thanks, @gustavosr8 and @ericonr. I've fixed all typos and implemented the minor suggestions.

> Considering the principle of each commit keeping the repository fully working, would it be possible to have "ioc: don't prune pre-selected directories." before "ioc: prune module directories from non-static builds."?

Conceptually, it makes sense. First, add the required tooling, then correctly prune things using them.

> Something like implement get_defined_paths_to_keep, then add it to relevant modules, then implement module pruning already using it? Would that make sense and be worth it?

As to whether it is worth it or not, I'm not sure. I'll give it a try and see if that's too much of a rework. At a first glance, it does not look like so, as "ioc: don't prune pre-selected directories." is pretty much self-contained and its usage is really simple. Refactoring messages is what will require a more careful look, I assume.

henriquesimoes commented:
Final diff is identical, as expected. I recommend using `git-range-diff(1)` for reviewing message changes ;)

Feel free to rebase and improve any of the messages at your will.
